src/providers/ldap/ldap_auth.c

```diff
@@ -695,10 +708,24 @@ static struct tevent_req *auth_connect_send(struct tevent_req *req)
                                "for debugging purposes only.");
     }
 
+    if (state->dn != NULL) {
+        /* In case the user's DN is known, the connection will only be used
+         * to bind as the user to perform the authentication. In that case,
+         * we don't need to authenticate the connection, because we're not
+         * looking up any information using the connection. This might be
+         * needed e.g. in case both ID and AUTH providers are set to LDAP
+         * and the server is AD, because otherwise the connection would
+         * both do a startTLS and later bind using GSSAPI which doesn't work
+         * well with AD.
+         */
+        need_authenticated_connection = false;
+    }
+
     subreq = sdap_cli_connect_send(state, state->ev, state->ctx->opts,
                                    state->ctx->be,
                                    state->sdap_service, false,
-                                   use_tls ? CON_TLS_ON : CON_TLS_OFF, false);
+                                   use_tls ? CON_TLS_ON : CON_TLS_OFF,
+                                   !need_authenticated_connection);
 
```

Comment:
I think it is a good idea to check if we already know the DN before the connection so that we might be able to avoid the authentication. However, if the DN is not know we end up with the same setup and first do TLS and later might do a GSSAPI/SASL bind to lookup the DN.

The use case would be e.g. local user which should authenticate with the AD password or a user with the same name and AD access is configured with GSSAPI. I didn't test this case yet, but I'm pretty sure it would fail as before.

So use_tls should only be true if authentication is not needed or if authentication is needed if authentication is done with default_bind_dn and password and not with GSSAPI/SASL bind.

btw, I would rename 'need_authenticated_connection' to 'skip_authentication' or similar so that the same logic as in sdap_cli_connect_send() is used and not the inverse one.

Reply:
Thank you for the review. I hope the new patch addresses your comments. I tested with these domain configurations:
```
[domain/win.trust.test]
#debug_level=9
cache_credentials = true
ldap_sasl_mech = GSSAPI
id_provider = ldap
ldap_schema = ad
ldap_uri = ldap://vagrant.win.trust.test
ldap_access_order = filter, expire
ldap_account_expire_policy = ad
ldap_id_mapping = true
ldap_search_base = dc=win,dc=trust,dc=test
dns_discovery_domain = win.trust.test
ldap_tls_reqcert=allow
default_shell = /bin/bash
fallback_homedir = /home/%u
enumerate=false
case_sensitive=false
chpass_provider=ldap
auth_provider=ldap
entry_cache_timeout=10800
krb5_realm = WIN.TRUST.TEST
timeout = 3000

[domain/adproxy]
id_provider = files

auth_provider = ldap
ldap_schema = ad
ldap_uri = ldap://vagrant.win.trust.test
ldap_search_base = dc=win,dc=trust,dc=test
debug_level = 10
timeout = 3000

ldap_default_bind_dn = CN=Administrator,CN=Users,DC=win,DC=trust,DC=test
ldap_default_authtok_type = password
ldap_default_authtok = ********

#ldap_sasl_mech = GSSAPI
#krb5_realm = WIN.TRUST.TEST

ldap_referrals = false
```

In the proxy domain I alternated between the password bind and the gssapi bind.